src/resolve/index.ts

```diff
@@ -6,17 +6,73 @@ import {Got} from '../network/got'
 import {Package} from '../types'
 import {LoggedPkg} from 'pnpm-logger'
 
-export type ResolveResult = {
+export type ResolutionBase = {
   id: string,
+}
+
+/**
+ * npm registry hosted package
+ */
+export type PackageResolution = ResolutionBase & {
+  type: 'package',
+  tarball: string,
+  shasum?: string,
   pkg?: Package,
+}
+
+/**
+ * tarball hosted remotely
+ */
+export type TarballResolution = ResolutionBase & {
+  type: 'tarball',
+  tarball: string,
+  shasum?: string,
+}
+
+/**
+ * tarball on a filesystem
+ */
+export type LocalTarballResolution = ResolutionBase & {
+  type: 'local-tarball',
+  tarball: string,
+  shasum?: string,
+}
```

Comment:
Probably we can merge `LocalTarballResolution` and `TarballResolution` together and differentiate by URL — if it starts with `file:` — it's a local one.